src/common/utils/validators.py

```diff
@@ -35,125 +33,157 @@ def is_blueprint_instance_of(
     return False
 
 
-def _recursive_validate_single_attribute(
-    attribute: BlueprintAttribute, value: dict | bool | int | float | str, get_blueprint: Callable, key: str
-):
-    debug_message = f"Location: Entity in key '{key}'"
-    if attribute.is_primitive:
-        python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
-        if not isinstance(value, python_type):
-            raise ValidationException(
-                f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
-                debug=debug_message,
-            )
-        if not attribute.is_optional:
-            if python_type == str and value == "":
-                raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=debug_message)
+def _validate_primitive_attribute(attribute: BlueprintAttribute, value: bool | int | float | str, key: str):
+    python_type = BuiltinDataTypes(attribute.attribute_type).to_py_type()
+    if type(value) != python_type:
+        raise ValidationException(
+            f"Attribute '{attribute.name}' should be type '{python_type.__name__}'. Got '{type(value).__name__}'",
+            debug=_get_debug_message(key),
+        )
+    if not attribute.is_optional and python_type == str and value == "":
+        raise ValidationException(f"Missing required attribute '{attribute.name}'", debug=_get_debug_message(key))
```

Comment:
Don't think we should raise an exception here.
Empty string is valid for a required string attribute

Reply:
Had to update the tests too, since we had a test which was to fail on empty strings